File: java/com/google/dotprompt/helpers/HelpersTest.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.dotprompt.helpers;

import static com.google.common.truth.Truth.assertThat;

import com.github.jknack.handlebars.EscapingStrategy;
import com.github.jknack.handlebars.Handlebars;
import com.github.jknack.handlebars.Template;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/**
 * Tests for the Helpers class.
 *
 * <p>These tests verify that all Handlebars helpers produce the expected output markers and handle
 * edge cases correctly. Test coverage matches the Python implementation for full parity.
 */
@RunWith(JUnit4.class)
public class HelpersTest {

  private Handlebars handlebars;

  @Before
  public void setUp() {
    handlebars = new Handlebars().with(EscapingStrategy.NOOP);
    Helpers.register(handlebars);
  }

  @Test
  public void testJsonHelper_basicObject() throws IOException {
    Map<String, Object> data = new HashMap<>();
    data.put("name", "John");
    data.put("age", 30);

    Template template = handlebars.compileInline("{{json data}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", data);

    String result = template.apply(context);
    assertThat(result).contains("\"name\":\"John\"");
    assertThat(result).contains("\"age\":30");
  }

  @Test
  public void testJsonHelper_withIndent() throws IOException {
    Map<String, Object> data = new HashMap<>();
    data.put("name", "John");
    data.put("age", 30);

    Template template = handlebars.compileInline("{{json data indent=2}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", data);

    // With indentation, the result should contain newlines
    String result = template.apply(context);
    assertThat(result).contains("\n");
    assertThat(result).contains("\"name\"");
    assertThat(result).contains("\"John\"");
  }

  @Test
  public void testJsonHelper_withArray() throws IOException {
    Template template = handlebars.compileInline("{{json data}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", Arrays.asList(1, 2, 3));

    String result = template.apply(context);
    assertThat(result).isEqualTo("[1,2,3]");
  }

  @Test
  public void testRoleHelper_system() throws IOException {
    Template template = handlebars.compileInline("{{role \"system\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:role:system>>>");
  }

  @Test
  public void testRoleHelper_user() throws IOException {
    Template template = handlebars.compileInline("{{role \"user\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:role:user>>>");
  }

  @Test
  public void testRoleHelper_model() throws IOException {
    Template template = handlebars.compileInline("{{role \"model\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:role:model>>>");
  }

  @Test
  public void testRoleHelper_empty() throws IOException {
    Template template = handlebars.compileInline("{{role}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("");
  }

  @Test
  public void testHistoryHelper() throws IOException {
    Template template = handlebars.compileInline("{{history}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:history>>>");
  }

  @Test
  public void testSectionHelper() throws IOException {
    Template template = handlebars.compileInline("{{section \"example\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:section example>>>");
  }

  @Test
  public void testSectionHelper_empty() throws IOException {
    Template template = handlebars.compileInline("{{section}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("");
  }

  @Test
  public void testMediaHelper_withUrl() throws IOException {
    Template template = handlebars.compileInline("{{media url=\"https://example.com/img.png\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:media:url https://example.com/img.png>>>");
  }

  @Test
  public void testMediaHelper_withUrlAndContentType() throws IOException {
    Template template =
        handlebars.compileInline(
            "{{media url=\"https://example.com/img.png\" contentType=\"image/png\"}}");
    String result = template.apply(new HashMap<>());
    assertThat(result).isEqualTo("<<<dotprompt:media:url https://example.com/img.png image/png>>>");
  }

  @Test
  public void testIfEquals_equalIntValues() throws IOException {
    Template template = handlebars.compileInline("{{#ifEquals a b}}yes{{else}}no{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 1);
    context.put("b", 1);
    assertThat(template.apply(context)).isEqualTo("yes");
  }

  @Test
  public void testIfEquals_unequalIntValues() throws IOException {
    Template template = handlebars.compileInline("{{#ifEquals a b}}yes{{else}}no{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 1);
    context.put("b", 2);
    assertThat(template.apply(context)).isEqualTo("no");
  }

  @Test
  public void testIfEquals_equalStringValues() throws IOException {
    Template template = handlebars.compileInline("{{#ifEquals a b}}yes{{else}}no{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", "test");
    context.put("b", "test");
    assertThat(template.apply(context)).isEqualTo("yes");
  }

  @Test
  public void testIfEquals_unequalStringValues() throws IOException {
    Template template = handlebars.compileInline("{{#ifEquals a b}}yes{{else}}no{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", "test");
    context.put("b", "diff");
    assertThat(template.apply(context)).isEqualTo("no");
  }

  @Test
  public void testUnlessEquals_unequalIntValues() throws IOException {
    Template template =
        handlebars.compileInline("{{#unlessEquals a b}}yes{{else}}no{{/unlessEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 1);
    context.put("b", 2);
    assertThat(template.apply(context)).isEqualTo("yes");
  }

  @Test
  public void testUnlessEquals_equalIntValues() throws IOException {
    Template template =
        handlebars.compileInline("{{#unlessEquals a b}}yes{{else}}no{{/unlessEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 1);
    context.put("b", 1);
    assertThat(template.apply(context)).isEqualTo("no");
  }

  @Test
  public void testUnlessEquals_unequalStringValues() throws IOException {
    Template template =
        handlebars.compileInline("{{#unlessEquals a b}}yes{{else}}no{{/unlessEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", "test");
    context.put("b", "diff");
    assertThat(template.apply(context)).isEqualTo("yes");
  }

  @Test
  public void testUnlessEquals_equalStringValues() throws IOException {
    Template template =
        handlebars.compileInline("{{#unlessEquals a b}}yes{{else}}no{{/unlessEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", "test");
    context.put("b", "test");
    assertThat(template.apply(context)).isEqualTo("no");
  }

  // Edge case tests for cross-runtime parity

  @Test
  public void testJsonHelper_with4SpaceIndent() throws IOException {
    Map<String, Object> data = new HashMap<>();
    data.put("test", true);

    Template template = handlebars.compileInline("{{json data indent=4}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", data);

    String result = template.apply(context);
    // Should have 4-space indentation
    assertThat(result).contains("    ");
    assertThat(result).contains("\n");
  }

  @Test
  public void testJsonHelper_nestedObjects() throws IOException {
    Map<String, Object> inner = new HashMap<>();
    inner.put("value", 42);
    Map<String, Object> middle = new HashMap<>();
    middle.put("inner", inner);
    Map<String, Object> outer = new HashMap<>();
    outer.put("outer", middle);

    Template template = handlebars.compileInline("{{json data}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", outer);

    String result = template.apply(context);
    assertThat(result).contains("\"outer\"");
    assertThat(result).contains("\"inner\"");
    assertThat(result).contains("\"value\":42");
  }

  @Test
  public void testJsonHelper_emptyObject() throws IOException {
    Template template = handlebars.compileInline("{{json data}}");
    Map<String, Object> context = new HashMap<>();
    context.put("data", new HashMap<>());

    String result = template.apply(context);
    assertThat(result).isEqualTo("{}");
  }

  @Test
  public void testIfEquals_typeSafety_intVsString() throws IOException {
    // Tests that 5 (Integer) != "5" (String) - strict type equality
    Template template =
        handlebars.compileInline("{{#ifEquals a b}}equal{{else}}not equal{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 5);
    context.put("b", "5");
    assertThat(template.apply(context)).isEqualTo("not equal");
  }

  @Test
  public void testIfEquals_booleanComparison_equal() throws IOException {
    Template template =
        handlebars.compileInline("{{#ifEquals a b}}equal{{else}}not equal{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", true);
    context.put("b", true);
    assertThat(template.apply(context)).isEqualTo("equal");
  }

  @Test
  public void testIfEquals_booleanComparison_unequal() throws IOException {
    Template template =
        handlebars.compileInline("{{#ifEquals a b}}equal{{else}}not equal{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", true);
    context.put("b", false);
    assertThat(template.apply(context)).isEqualTo("not equal");
  }

  @Test
  public void testIfEquals_nullComparison_bothNull() throws IOException {
    Template template =
        handlebars.compileInline("{{#ifEquals a b}}equal{{else}}not equal{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", null);
    context.put("b", null);
    assertThat(template.apply(context)).isEqualTo("equal");
  }

  @Test
  public void testIfEquals_nullComparison_nullVsZero() throws IOException {
    Template template =
        handlebars.compileInline("{{#ifEquals a b}}equal{{else}}not equal{{/ifEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", null);
    context.put("b", 0);
    assertThat(template.apply(context)).isEqualTo("not equal");
  }

  @Test
  public void testUnlessEquals_typeSafety_intVsString() throws IOException {
    // Tests that 5 (Integer) != "5" (String) - strict type inequality
    Template template =
        handlebars.compileInline("{{#unlessEquals a b}}not equal{{else}}equal{{/unlessEquals}}");

    Map<String, Object> context = new HashMap<>();
    context.put("a", 5);
    context.put("b", "5");
    assertThat(template.apply(context)).isEqualTo("not equal");
  }
}

File: js/src/helpers.test.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { describe, expect, it } from 'vitest';
import {
  history,
  ifEquals,
  json,
  media,
  role,
  section,
  unlessEquals,
} from './helpers';

describe('helpers', () => {
  describe('json', () => {
    it('should serialize an object to JSON', () => {
      const result = json({ foo: 'bar' }, { hash: {} });
      expect(result.toString()).toBe('{"foo":"bar"}');
    });

    it('should serialize with indent when specified', () => {
      const result = json({ foo: 'bar' }, { hash: { indent: 2 } });
      expect(result.toString()).toBe('{\n  "foo": "bar"\n}');
    });

    it('should serialize with 4-space indent when specified', () => {
      const result = json({ test: true }, { hash: { indent: 4 } });
      expect(result.toString()).toBe('{\n    "test": true\n}');
    });

    it('should serialize nested objects', () => {
      const result = json({ outer: { inner: { value: 42 } } }, { hash: {} });
      expect(result.toString()).toBe('{"outer":{"inner":{"value":42}}}');
    });

    it('should serialize arrays', () => {
      const result = json([1, 2, 3], { hash: {} });
      expect(result.toString()).toBe('[1,2,3]');
    });

    it('should serialize empty objects', () => {
      const result = json({}, { hash: {} });
      expect(result.toString()).toBe('{}');
    });

    it('should serialize empty arrays', () => {
      const result = json([], { hash: {} });
      expect(result.toString()).toBe('[]');
    });

    it('should throw on circular references (fail-fast behavior)', () => {
      const circular: Record<string, unknown> = { foo: 'bar' };
      circular.self = circular;

      expect(() => json(circular, { hash: {} })).toThrow(TypeError);
    });

    it('should throw on BigInt values (fail-fast behavior)', () => {
      // BigInt is not serializable to JSON
      expect(() => json({ value: BigInt(123) }, { hash: {} })).toThrow(
        TypeError
      );
    });
  });

  describe('role', () => {
    it('should return a role marker for system', () => {
      const result = role('system');
      expect(result.toString()).toBe('<<<dotprompt:role:system>>>');
    });

    it('should return a role marker for user', () => {
      const result = role('user');
      expect(result.toString()).toBe('<<<dotprompt:role:user>>>');
    });

    it('should return a role marker for model', () => {
      const result = role('model');
      expect(result.toString()).toBe('<<<dotprompt:role:model>>>');
    });
  });

  describe('history', () => {
    it('should return a history marker', () => {
      const result = history();
      expect(result.toString()).toBe('<<<dotprompt:history>>>');
    });
  });

  describe('section', () => {
    it('should return a section marker', () => {
      const result = section('examples');
      expect(result.toString()).toBe('<<<dotprompt:section examples>>>');
    });
  });

  describe('media', () => {
    it('should return a media marker with url only', () => {
      const result = media({
        hash: { url: 'https://example.com/image.png' },
      } as Handlebars.HelperOptions);
      expect(result.toString()).toBe(
        '<<<dotprompt:media:url https://example.com/image.png>>>'
      );
    });

    it('should return a media marker with url and contentType', () => {
      const result = media({
        hash: {
          url: 'https://example.com/image.png',
          contentType: 'image/png',
        },
      } as Handlebars.HelperOptions);
      expect(result.toString()).toBe(
        '<<<dotprompt:media:url https://example.com/image.png image/png>>>'
      );
    });
  });

  describe('ifEquals', () => {
    const mockContext = {};
    const createOptions = (fn: () => string, inverse: () => string) =>
      ({
        fn: () => fn(),
        inverse: () => inverse(),
      }) as unknown as Handlebars.HelperOptions;

    it('should return fn result when values are equal', () => {
      const options = createOptions(
        () => 'equal',
        () => 'not equal'
      );
      const result = ifEquals.call(mockContext, 5, 5, options);
      expect(result).toBe('equal');
    });

    it('should return inverse result when values are not equal', () => {
      const options = createOptions(
        () => 'equal',
        () => 'not equal'
      );
      const result = ifEquals.call(mockContext, 5, 6, options);
      expect(result).toBe('not equal');
    });

    it('should use strict equality (different types are not equal)', () => {
      const options = createOptions(
        () => 'equal',
        () => 'not equal'
      );
      // 5 !== "5" in strict equality
      const result = ifEquals.call(mockContext, 5, '5', options);
      expect(result).toBe('not equal');
    });

    it('should compare boolean values correctly', () => {
      const options = createOptions(
        () => 'equal',
        () => 'not equal'
      );
      expect(ifEquals.call(mockContext, true, true, options)).toBe('equal');
      expect(ifEquals.call(mockContext, false, false, options)).toBe('equal');
      expect(ifEquals.call(mockContext, true, false, options)).toBe(
        'not equal'
      );
    });

    it('should compare null values correctly', () => {
      const options = createOptions(
        () => 'equal',
        () => 'not equal'
      );
      expect(ifEquals.call(mockContext, null, null, options)).toBe('equal');
      expect(ifEquals.call(mockContext, null, 0, options)).toBe('not equal');
      expect(ifEquals.call(mockContext, null, undefined, options)).toBe(
        'not equal'
      );
    });
  });

  describe('unlessEquals', () => {
    const mockContext = {};
    const createOptions = (fn: () => string, inverse: () => string) =>
      ({
        fn: () => fn(),
        inverse: () => inverse(),
      }) as unknown as Handlebars.HelperOptions;

    it('should return fn result when values are not equal', () => {
      const options = createOptions(
        () => 'not equal',
        () => 'equal'
      );
      const result = unlessEquals.call(mockContext, 5, 6, options);
      expect(result).toBe('not equal');
    });

    it('should return inverse result when values are equal', () => {
      const options = createOptions(
        () => 'not equal',
        () => 'equal'
      );
      const result = unlessEquals.call(mockContext, 5, 5, options);
      expect(result).toBe('equal');
    });

    it('should use strict inequality (different types are not equal)', () => {
      const options = createOptions(
        () => 'not equal',
        () => 'equal'
      );
      // 5 !== "5" in strict comparison
      const result = unlessEquals.call(mockContext, 5, '5', options);
      expect(result).toBe('not equal');
    });

    it('should compare boolean values correctly', () => {
      const options = createOptions(
        () => 'not equal',
        () => 'equal'
      );
      expect(unlessEquals.call(mockContext, true, false, options)).toBe(
        'not equal'
      );
      expect(unlessEquals.call(mockContext, true, true, options)).toBe('equal');
    });

    it('should compare null values correctly', () => {
      const options = createOptions(
        () => 'not equal',
        () => 'equal'
      );
      expect(unlessEquals.call(mockContext, null, null, options)).toBe('equal');
      expect(unlessEquals.call(mockContext, null, 0, options)).toBe(
        'not equal'
      );
    });
  });
});
